File: npai-agent/client/npaiagent-bench.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# __author__ = "Wilson Lan"

import socket
import time
import sys
import os
import argparse
import errno
import multiprocessing

# NOTE:
#   In some cases, cgroups.py doesn't locate in current directory.
#   Callers should be aware of this situation and expand the syspath, 
#   e.g. sys.path.append("your_own_python_lib_path")
sys.path.append("/home/lanxinyu/nats/pylibs")
import cgroups

DEFAULT_NUM_MSGS      = 1000000
DEFAULT_MSG_SIZE      = 256
DEFAULT_NUM_PROCESSES = 2


def show_usage():
    message = """
Usage: npaiagent_test [options]

options:
    -n, --count     COUNT              Messages to send (default: 1000000)
    -s, --size      SIZE               Message size (default: 256)
    -u, --url       URL                NPAI agent address (default: /var/run/npai-agent.sock)
    -c, --core      CORE               Set CPU affinity (default: all available CPUs)
    -p, --processes NUM                Number of subprocess to send messages (default: 2)
    """
    print(message)

def show_usage_and_die():
    show_usage()
    sys.exit(1)

def configure_cgroups(core):
    cg = []
    cg.append(cgroups.must_make_cgroup("cpuset", "npaiagent_test"))
    cg.append(cgroups.must_make_cgroup("cpu", "npaiagent_test"))
    
    # set cpuset attributes
    if core == "":
        cgroups.set_cgroups_cpuset_attr(*cg)
    else:
        cgroups.set_cgroups_cpuset_attr(*cg, core_id=core)

    # set cgroup process
    cgroups.set_cgroups_procs(os.getpid(), *cg)

    return cg

def send_message(url, count, size):
    # Initialize a unix socket
    sock = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)

    # Connect to npai-agent 
    returncode = sock.connect_ex(url)
    if returncode != 0:
        print "Error: failed to connect {url}: {errmsg}".format(
                url=url, errmsg=os.strerror(returncode))
        return

    # Formulate packet payload
    data = []
    for i in range(0, size):
        data.append(b'W')
    data.append(b'\n')

    payload = b''.join(data)

    # Deliver messages
    for i in xrange(count):
        # NOTE Using `sendall` will lead to poor performance.
        # Perhaps, the calling of `send` over unix socket is okay.
        # However, we should write like this in those serious occasions:
        ## sent = 0
        ## while sent < size+1:
        ##     nbytes = sock.send(payload[sent:])
        ##     sent += nbytes
        sock.send(payload)

    sock.close()

def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("-n", "--count", default=DEFAULT_NUM_MSGS, type=int)
    parser.add_argument("-s", "--size", default=DEFAULT_MSG_SIZE, type=int)
    parser.add_argument("-u", "--url", default="/var/run/npai-agent.sock")
    parser.add_argument("-c", "--core", default="")
    parser.add_argument("-p", "--processes", default=DEFAULT_NUM_PROCESSES,
            type=int)
    args = parser.parse_args()

    # Configure Cgroups
    cg = configure_cgroups(args.core)

    # Start the benchmark
    start = time.time()
    p = multiprocessing.Pool()
    for i in xrange(args.processes):
        p.apply_async(send_message, args=(args.url, 
            args.count/args.processes, args.size))
    p.close()
    p.join()
    elapsed = time.time() - start

    # Print throughput
    throughput = args.count / elapsed
    rate = args.size * args.count * 8 / elapsed / 1e6
    print "Send %d messages (size %d) in %.2f seconds" % \
        (args.count, args.size, elapsed)
    print "Rate %.2f msg/s, %.2f Mbps" % (throughput, rate)

    # Clean Cgroups
    cgroups.clean_cgroups(*cg)


if __name__ == "__main__":
    main()

File: npai-agent/client/npaiagent-bench.go
package main

import (
	"flag"
	"fmt"
	"log"
	"net"
	"os"
	"runtime"
	"time"

	"github.com/containerd/cgroups"
	"github.com/nats-io/mq-benchmarking/benchmark"
	specs "github.com/opencontainers/runtime-spec/specs-go"
)

const (
	DefaultUnixSockPath = "/var/run/npai-agent.sock"
	DefaultMessageCount = 1000000
	DefaultMessageSize  = 1000
)

var usageStr = `
Usage npaiagent_test [options]

Options:
    -u, --url  <url>           Npai agent URL (default: /var/run/npai-agent.sock)
    -c, --core <string>        Set CPU afinity, eg: 1|2-3 (default: all available CPUs)
    -n, --num_messages <int>   Number of total messages (default: 1000000)
    -s, --message_size <int>   Size of message (default: 1000)
    -r, --random               Send random bytes instead of empty bytes (default: false)
`

func usage() {
	log.Fatalf(usageStr)
}

func main() {
	var url string
	var core string
	var messageCount int
	var messageSize int
	var random bool
	flag.StringVar(&url, "u", DefaultUnixSockPath, "Npai agent URL")
	flag.StringVar(&url, "url", DefaultUnixSockPath, "Npai agent URL")
	flag.StringVar(&core, "c", "", "Set CPU afinity")
	flag.StringVar(&core, "core", "", "Set CPU afinity")
	flag.IntVar(&messageCount, "n", DefaultMessageCount, "Number of total messages")
	flag.IntVar(&messageCount, "num_messages", DefaultMessageCount, "Number of total messages")
	flag.IntVar(&messageSize, "s", DefaultMessageSize, "Size of message")
	flag.IntVar(&messageSize, "message_size", DefaultMessageSize, "Size of message")
	flag.BoolVar(&random, "r", false, "Send random bytes instead of empty bytes")
	flag.BoolVar(&random, "random", false, "Send random bytes instead of empty bytes")
	log.SetFlags(0)
	flag.Usage = usage
	flag.Parse()

	// Configure Cgroups
	memNode := "0"
	if core == "" {
		core = fmt.Sprintf("0-%d", runtime.NumCPU()-1)
	}
	cgroupName := "/npaiagent_test"
	cgCtrl, err := cgroups.New(cgroups.V1, cgroups.StaticPath(cgroupName),
		&specs.LinuxResources{
			CPU: &specs.LinuxCPU{
				Cpus: core,
				Mems: memNode,
			},
		})
	if err != nil {
		log.Fatalf("Error: failed to create cgroup: %v", err)
	}
	if err := cgCtrl.Add(cgroups.Process{Pid: os.Getpid()}); err != nil {
		log.Fatalf("Error: failed to add process to cgroup: %v", err)
	}

	// Connect to npai-agent
	conn, err := net.Dial("unix", url)
	if err != nil {
		log.Fatalf("Error: Can't connect to npai-agent: %v", err)
	}

	// Formulate payload
	message := make([]byte, messageSize)
	if random {
		benchmark.RandBytes(message)
	}
	message = append(message, '\n')

	// Start benchmarking
	start := time.Now().UnixNano()
	for i := 0; i < messageCount; i++ {
		conn.Write(message)
	}
	stop := time.Now().UnixNano()
	elapsed := float32(stop-start) / 1e9
	log.Printf("Sent %d messages (size %d) in %.2f s",
		messageCount,
		messageSize,
		elapsed)
	log.Printf("Rate %.2f Mbps, %.2f msg/s",
		float32(8*messageSize*messageCount)/elapsed/1e6,
		float32(messageCount)/elapsed)

	// Close connection
	conn.Close()

	// Clean Cgroups
	rootCgCtrl, err := cgroups.Load(cgroups.V1, cgroups.RootPath)
	if err != nil {
		log.Fatalf("Error: failed to load root cgroup: %v", err)
	}
	cgCtrl.MoveTo(rootCgCtrl)
	cgCtrl.Delete()
}
